### PR TITLE
Feat/deposit

### PR DESCRIPTION
**Pull request by LHFVII**
## Summary by CodeRabbit

* **New Features**
  * Deposit flow: users can deposit assets, receive minted shares, specify a fee recipient, and benefit from mint-authority reassignment and configurable rounding for share calculations.
* **Bug Fixes**
  * Improved arithmetic safety and new user-facing errors for paused vaults, asset caps, insufficient deposits, and mint/supply issues.
* **Tests**
  * Added extensive integration tests covering deposits, fees, transfer-fee scenarios, and lifecycle flows.
* **Documentation**
  * Expanded docs for vault lifecycle, deposit behavior, fees, authority, and update parameters.

### CODE REVIEW

File: programs/vault/src/instructions/deposit.rs [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        .total_asset_balance
+        .checked_add(assets)
+        .ok_or(VaultProgramError::ArithmeticError)?
+        .checked_sub(fee)
+        .ok_or(VaultProgramError::ArithmeticError)?;
+
+    require!(
+        expected_new_total_asset_balance <= ctx.accounts.vault.vault_asset_cap,
+        VaultProgramError::MaxVaultAssetCapExceeded
+    );
```
[the remark this answers is not in the document]
Comment (LHFVII) [reply programs/vault/src/instructions/deposit.rs]: Does not apply it can be updated

File: integration-tests/src/vault/deposit.rs [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        asset_mint.pubkey(),
+        share_mint.pubkey(),
+        reserve_pubkey,
+        vault_pubkey,
+        deposit_fee.clone(),
+        withdraw_fee.clone(),
+        0,
+        100_000,
+        fee_recipient.pubkey(),
+    );
```
[the remark this answers is not in the document]
Comment (LHFVII) [reply integration-tests/src/vault/deposit.rs]: N/A

File: integration-tests/src/vault/deposit.rs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        &authority,
+        asset_mint.pubkey(),
+        share_mint.pubkey(),
+        vault_pubkey,
+        updated_deposit_fee.clone(),
+        updated_withdraw_fee.clone(),
+        updated_vault_asset_cap,
+        updated_paused_status,
+        authority.pubkey(),
+    );
```
[the remark this answers is not in the document]
Comment (LHFVII) [reply integration-tests/src/vault/deposit.rs]: N/A

File: programs/vault/src/lib.rs [hunk 4]
```diff
@@ -12,18 +12,52 @@ pub mod vault {
...
+    /// * `fee_recipient` - The pubkey that will receive collected fees
     pub fn create_vault(ctx: Context<CreateVault>, args: VaultArgs) -> Result<()> {
         instructions::create_vault::handler(ctx, args)
     }
 
-    /// Closes an atomic vault.
+    /// Closes a vault and transfers all remaining assets to a specified destination.
+    /// Transfers any remaining reserve assets to the provided destination token account,
+    /// closes the reserve token account, and closes the vault config account.
+    /// Only the vault authority can close the vault.
```
[the remark this answers is not in the document]
Comment (LHFVII) [reply programs/vault/src/lib.rs]: N/A

File: programs/vault/src/state/vault_config.rs [hunk 5]
```diff
@@ -48,6 +65,65 @@ pub struct VaultConfig {
...
     pub reserve_bump: u8,
     pub bump: u8,
 }
+
+impl VaultConfig {
+    pub fn total_assets(self) -> u64 {
+        return self.total_asset_balance;
+    }
+
+    pub fn get_shares_from_assets(
```
Comment (sebatustra) [inline programs/vault/src/state/vault_config.rs:78]: I'm a bit confused by the function name, shouldn't `share_amount` be named `asset_amount`? That way it's actually assets -> shares

File: integration-tests/src/vault/deposit.rs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    assert_eq!(result.is_ok(), true, "Unexpected result for test case");
+
+    fee_recipient_ata_account = svm
+        .get_account(&fee_recipient_ata)
+        .expect("Vault account should exist");
+
+    let fee_recipient_balance_after = TokenAccount::unpack(fee_recipient_ata_account.data())
+        .unwrap()
+        .amount;
```
Comment (Taylor123) [inline integration-tests/src/vault/deposit.rs:119]: probably worth having some utils for "get_token_account" and the like from LiteSVM. Could be a good additional to our repository template as well

File: integration-tests/src/vault/deposit.rs [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        .expect("Vault account should exist");
+
+    let user_share_balance_after = TokenAccount::unpack(user_share_ata_account.data())
+        .unwrap()
+        .amount;
+
+    assert_eq!(
+        user_share_balance_after,
+        deposit_amount.checked_sub(fee).expect("overflow")
+    );
```
Comment (Taylor123) [inline integration-tests/src/vault/deposit.rs:363]: missing some assertions. Some that come to mind:
- share mint supply value
- Vault's TokenAccount balance

**Review by Taylor123 — COMMENTED** [review]
As general feedback, it would be good to see more comments and comments with more detail. This is incredibly helpful for auditors to understand the intention of an instruction. If the comment and instruction logic diverge, auditors/maintainers can then flag it as a potential problem or make sure that the program still works as design intended.

File: programs/vault/src/instructions/create_vault.rs [hunk 8]
```diff
@@ -1,29 +1,31 @@
...
     pub payer: Signer<'info>,
 
-    #[account()]
+    pub mint_authority: Signer<'info>,
+
     pub asset_mint: InterfaceAccount<'info, Mint>,
 
-    #[account()]
+    #[account(mut)]
     pub share_mint: InterfaceAccount<'info, Mint>,
```
Comment (Taylor123) [inline programs/vault/src/instructions/create_vault.rs:32]: if we're not going to check the mint authority matches (fine if we don't), then let's add a comment that indicates this is intended. Otherwise it could get flagged during audit

File: programs/vault/src/instructions/deposit.rs [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    #[account(
+        mut,
+        associated_token::authority = user,
+        associated_token::mint = share_mint,
+        associated_token::token_program = token_program,
+    )]
+    pub user_shares_account: InterfaceAccount<'info, TokenAccount>,
+
+    pub token_program: Interface<'info, TokenInterface>,
+    pub reserve_token_program: Interface<'info, TokenInterface>,
```
Comment (Taylor123) [inline programs/vault/src/instructions/deposit.rs]: nit: when using two token programs due to different mints, IMO it's a better convention to name it based on what you name the mints. "token_program" does not indicate any association.

Updating to "asset_token_program" and "share_token_program" makes it much more clear which is which IMO

File: programs/vault/src/instructions/deposit.rs [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    pub associated_token_program: Program<'info, AssociatedToken>,
+    pub system_program: Program<'info, System>,
+}
+
+impl<'info> Deposit<'info> {
+    pub fn transfer_reserve_token_fee_to_fee_recipient(&mut self, fee: u64) -> Result<()> {
```
Comment (Taylor123) [inline programs/vault/src/instructions/deposit.rs]: nit: let's stick to "asset" and "share" naming conventions. Reserve refers to a TokenAccount, not a mint. So "transfer_reserve_token" should really be "transfer_asset_token" IMO since it's more clear that it refers to the Mint and not a TokenAccount

File: programs/vault/src/instructions/deposit.rs [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        };
+        let cpi_ctx = CpiContext::new(
+            self.reserve_token_program.to_account_info(),
+            fee_recipient_transfer_cpi_accounts,
+        );
+
+        token_interface::transfer_checked(cpi_ctx, fee, self.asset_mint.decimals)
+    }
+
+    pub fn transfer_reserve_token_to_vault(&mut self, amount: u64) -> Result<()> {
```
Comment (Taylor123) [inline programs/vault/src/instructions/deposit.rs]: same naming convention update needed

File: programs/vault/src/instructions/deposit.rs [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        };
+
+        let cpi_ctx = CpiContext::new(
+            self.reserve_token_program.to_account_info(),
+            vault_transfer_cpi_accounts,
+        );
+        token_interface::transfer_checked(cpi_ctx, amount, self.asset_mint.decimals)
+    }
+
+    pub fn mint(&mut self, amount: u64) -> Result<()> {
```
Comment (Taylor123) [inline programs/vault/src/instructions/deposit.rs]: nit: naming should be more explicit to indicate intention. Would name to something like "mint_shares_to_user"

File: programs/vault/src/instructions/deposit.rs [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        .total_asset_balance
+        .checked_add(assets)
+        .ok_or(VaultProgramError::ArithmeticError)?
+        .checked_sub(fee)
+        .ok_or(VaultProgramError::ArithmeticError)?;
+
+    require!(
+        expected_new_total_asset_balance <= ctx.accounts.vault.vault_asset_cap,
+        VaultProgramError::MaxVaultAssetCapExceeded
+    );
```
Comment (Taylor123) [inline programs/vault/src/instructions/deposit.rs]: this does not work for Token2022 Mints that have TransferFees. IMO it should likely perform the token transfer and calculate the balance difference from the transfer and use that to perform the check.

We should also add a test case for this.

File: programs/vault/src/instructions/deposit.rs [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+
+    let remaining_amount = assets
+        .checked_sub(fee)
+        .ok_or(VaultProgramError::ArithmeticError)?;
+
+    let shares = ctx.accounts.vault.get_shares_from_assets(
+        &ctx.accounts.share_mint,
+        remaining_amount,
+        Rounding::Down,
+    )?;
```
Comment (Taylor123) [inline programs/vault/src/instructions/deposit.rs]: Similar to the above, this also does not work for Token2022. We should perform the asset transfer to the vault, get the diff, and then use that to calculate the amount of shares that need to be minted

File: programs/vault/src/state/vault_config.rs [hunk 15]
```diff
@@ -22,12 +22,29 @@ impl FeeType {
...
+                    .checked_add(9_999)
+                    .ok_or(VaultProgramError::ArithmeticError)?
+                    .checked_div(10_000)
+                    .ok_or(VaultProgramError::ArithmeticError)?;
+                return Ok(fee);
+            }
+            FeeType::FixedAmount { amount } => return Ok(amount),
+            FeeType::NoFee => return Ok(0),
+        }
+    }
```
Comment (Taylor123) [inline programs/vault/src/state/vault_config.rs:40]: Let's add unit tests for this please

File: programs/vault/src/state/vault_config.rs (hunk 5 above)
Comment (Taylor123) [inline programs/vault/src/state/vault_config.rs:78]: This function definitely needs unit tests given it's a critical piece of logic

Comment (Taylor123) [reply programs/vault/src/state/vault_config.rs:78]: should make sure we test various edge cases and all combinations. For example:
- share_mint.supply == 0 && self.total_asset_balance == 0
- share_mint.supply > 0 && self.total_asset_balance == 0
- share_mint.supply == 0 && self.total_asset_balance > 0
- share_mint.supply > 0 && self.total_asset_balance > 0
- share_mint.supply == u64::MAX
- self.total_asset_balance == u64::MAX

File: programs/vault/src/state/vault_config.rs [hunk 16]
```diff
@@ -48,6 +65,275 @@ pub struct VaultConfig {
...
+        self,
+        supply: u64,
+        asset_amount: u64,
+        rounding: Rounding,
+    ) -> Result<u64> {
+        let mut assets_times_total_supply: u128;
+        if supply == 0 {
+            assets_times_total_supply = u128::from(self.initial_price)
+                .checked_mul(u128::from(asset_amount))
+                .ok_or(VaultProgramError::ArithmeticError)?;
```
[the remark this answers is not in the document]
Comment (LHFVII) [reply programs/vault/src/state/vault_config.rs]: Does not apply

File: programs/vault/src/instructions/deposit.rs [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+    // Not checking the mint authority is expected behaviour
+    #[account(mut)]
+    pub share_mint: InterfaceAccount<'info, Mint>,
+
+    #[account(
+        mut,
+        seeds = [RESERVE_CONFIG_SEED, asset_mint.key().as_ref(), share_mint.key().as_ref()],
+        bump,
+    )]
+    pub reserve: InterfaceAccount<'info, TokenAccount>,
```
[the remark this answers is not in the document]
Comment (LHFVII) [reply programs/vault/src/instructions/deposit.rs:31]: Fixed

File: integration-tests/src/vault/deposit.rs [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+
+    let mint_account = spl_token::state::Mint::unpack(&share_mint_account.data).unwrap();
+    assert_eq!(
+        mint_account.supply,
+        deposit_amount
+            .checked_sub(fee)
+            .expect("overflow")
+            .checked_sub(transfer_fee_deposit_amount)
+            .expect("overflow")
+    );
```
Comment (Taylor123) [inline integration-tests/src/vault/deposit.rs:363]: i think this is missing the assertion for the virtual vault balance, which is really important. The vault virtual balance could still using logic with args.amount and this test would continue to pass

Comment (Taylor123) [reply integration-tests/src/vault/deposit.rs:363]: important to make sure we are asserting against all state changes (not just token transfers/supply)

File: programs/vault/src/instructions/deposit.rs [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+            seeds,
+        );
+        mint_to(mint_cpi_ctx, amount)
+    }
+}
+pub fn handler<'info>(ctx: Context<Deposit>, assets: u64) -> Result<()> {
+    require!(!ctx.accounts.vault.paused, VaultProgramError::PausedVault);
+    let fee = ctx.accounts.vault.get_deposit_fee(assets)?;
+    // current vault amount
+    let current_reserve_amount = ctx.accounts.reserve.amount;
```
Comment (Taylor123) [inline programs/vault/src/instructions/deposit.rs]: nit: i'd prefer this to be pre_reserve_amount or reserve_amount_before since when this is used after the transfer happens "current" is no longer applicable

File: programs/vault/src/instructions/deposit.rs [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+        .transfer_asset_token_to_vault(remaining_amount)?;
+    ctx.accounts.reserve.reload()?;
+
+    let updated_reserve_amount = ctx.accounts.reserve.amount;
+
+    let actual_transferred_amount = updated_reserve_amount
+        .checked_sub(current_reserve_amount)
+        .ok_or(VaultProgramError::ArithmeticError)?;
+
+    let expected_new_total_asset_balance = ctx
```
Comment (Taylor123) [inline programs/vault/src/instructions/deposit.rs]: nit: this is no longer "expected", it is the new_total_asset_balance

File: programs/vault/src/state/vault_config.rs [hunk 21]
```diff
@@ -48,6 +65,275 @@ pub struct VaultConfig {
...
+            withdraw_fees: FeeType::NoFee,
+            paused: false,
+            vault_asset_cap: u64::MAX,
+            total_asset_balance,
+            fee_recipient: Pubkey::new_unique(),
+            reserve_bump: 255,
+            bump: 254,
+        }
+    }
+
```
Comment (Taylor123) [inline programs/vault/src/state/vault_config.rs:150]: could use test_case with parameters if we want to DRY up some of the code

File: programs/vault/src/state/vault_config.rs [hunk 22]
```diff
@@ -48,6 +65,275 @@ pub struct VaultConfig {
...
+        return self.total_asset_balance;
+    }
+
+    pub fn get_shares_from_assets(
+        self,
+        supply: u64,
+        asset_amount: u64,
+        rounding: Rounding,
+    ) -> Result<u64> {
+        let mut assets_times_total_supply: u128;
```
Comment (Taylor123) [inline programs/vault/src/state/vault_config.rs]: super nite: could just have the if/else statement return a value so then this does not have to be mutable

File: programs/vault/src/state/vault_config.rs (hunk 15 above)
Comment (Taylor123) [reply programs/vault/src/state/vault_config.rs:40]: @LHFVII could we add some minor unit testing for this function as well

**Review by sebatustra — APPROVED** [review]
LGTM

File: programs/vault/src/state/vault_config.rs [hunk 23]
```diff
@@ -48,6 +64,192 @@ pub struct VaultConfig {
...
+            Rounding::Down => assets_times_total_supply
+                .checked_div(u128::from(
+                    self.total_assets()
+                        .checked_add(1)
+                        .ok_or(VaultProgramError::ArithmeticError)?,
+                ))
+                .ok_or(VaultProgramError::ArithmeticError)?,
+        };
+        u64::try_from(result).or(Err(VaultProgramError::ArithmeticError.into()))
+    }
```
[the remark this answers is not in the document]
Comment (LHFVII) [reply programs/vault/src/state/vault_config.rs:112]: Applied
